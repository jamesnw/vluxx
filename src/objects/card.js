export default class Card {
  constructor(title, type, apply) {
    this.title = title;
    this.type = type;
    this.apply = apply;
  }
}
