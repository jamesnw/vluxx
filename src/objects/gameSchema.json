{
  "type": "object",
  "definitions": {
    "card": {
      "type": "object",
      "properties": {
        "title": { "type": "string" },
        "type": { "type": "string" }
      },
      "required": ["title", "type"]
    },
    "player": {
      "type": "object",
      "properties": {
        "hand": {
          "type": "array",
          "items": [
            {
              "$ref": "#/definitions/card"
            }
          ]
        },
        "name": {
          "type": "string"
        }
      },
      "required": ["hand", "name"]
    }
  },
  "properties": {
    "activePlayer": {
      "type": "integer"
    },
    "constantRules": {
      "type": "object",
      "properties": {
        "draw": {
          "type": "integer"
        },
        "play": {
          "type": "integer"
        }
      },
      "required": ["draw", "play"]
    },
    "deck": {
      "type": "array",
      "items": [
        {
          "$ref": "#/definitions/card"
        }
      ]
    },
    "discard": {
      "type": "array",
      "items": [
        {
          "$ref": "#/definitions/card"
        }
      ]
    },
    "name": {
      "type": "string"
    },
    "players": {
      "type": "array",
      "items": [{ "$ref": "#/definitions/player" }]
    },
    "rules": {
      "type": "array",
      "items": {}
    },
    "state": {
      "type": "string",
      "enum": ["started", "won"]
    },
    "turn": {
      "type": "object",
      "properties": {
        "playsRemaining": {
          "type": ["integer", "null"]
        }
      },
      "required": ["playsRemaining"]
    }
  },
  "required": [
    "activePlayer",
    "constantRules",
    "deck",
    "discard",
    "name",
    "players",
    "rules",
    "state",
    "turn"
  ]
}
