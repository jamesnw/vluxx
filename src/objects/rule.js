export default class Rule {
  constructor(name) {
    this.name = name;
  }
}
