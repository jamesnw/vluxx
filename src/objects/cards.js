import { gameIsValid, checkForWinner } from "./game";
const TYPES = {
  GOAL: "goal",
  RULE: "rule",
  KEEPER: "keeper"
};
export const cards = [
  {
    title: "Chocolate Chip Cookie",
    type: TYPES.GOAL,
    goal: ["Chocolate", "Cookie"]
  },
  { title: "Bedtime snack", type: TYPES.GOAL, goal: ["Milk", "Cookie"] },
  { title: "Chocolate Stout", type: TYPES.GOAL, goal: ["Chocolate", "Beer"] },
  {
    title: "Chocolate covered pretzel",
    type: TYPES.GOAL,
    goal: ["Chocoloate", "Pretzel"]
  },
  { title: "Orange Julius", type: TYPES.GOAL, goal: ["Milk", "Orange"] },
  {
    title: "Just Missing Vodka",
    type: TYPES.GOAL,
    goal: ["Screwdriver", "Orange"]
  },
  { title: "Tools", type: TYPES.GOAL, goal: ["Hammer", "Screwdriver"] },

  { title: "Chocolate", type: TYPES.KEEPER },
  { title: "Cookie", type: TYPES.KEEPER },
  { title: "Milk", type: TYPES.KEEPER },
  { title: "Beer", type: TYPES.KEEPER },
  { title: "Orange", type: TYPES.KEEPER },
  { title: "Book", type: TYPES.KEEPER },
  { title: "Pretzel", type: TYPES.KEEPER },
  { title: "Hammer", type: TYPES.KEEPER },
  { title: "Screwdriver", type: TYPES.KEEPER },
  { title: "Milk7", type: TYPES.KEEPER },
  { title: "Milk8", type: TYPES.KEEPER },
  { title: "Milk9", type: TYPES.KEEPER },
  { title: "Milk0", type: TYPES.KEEPER },
  { title: "Milk12", type: TYPES.KEEPER },
  { title: "Milk23", type: TYPES.KEEPER },
  { title: "Milk24", type: TYPES.KEEPER },
  { title: "Milk35", type: TYPES.KEEPER },

  { title: "Draw 2", type: TYPES.RULE, rule: { category: "Draw", amount: 2 } },
  { title: "Draw 3", type: TYPES.RULE, rule: { category: "Draw", amount: 3 } },
  { title: "Draw 4", type: TYPES.RULE, rule: { category: "Draw", amount: 4 } },
  { title: "Draw 5", type: TYPES.RULE, rule: { category: "Draw", amount: 5 } },
  { title: "Play 2", type: TYPES.RULE, rule: { category: "Play", amount: 2 } },
  { title: "Play 3", type: TYPES.RULE, rule: { category: "Play", amount: 3 } },
  { title: "Play 4", type: TYPES.RULE, rule: { category: "Play", amount: 4 } },
  { title: "Play 5", type: TYPES.RULE, rule: { category: "Play", amount: 5 } }
];

export function applyCard(game, card) {
  switch (card.type) {
    case TYPES.GOAL:
      game.goal = card;
      break;
    case TYPES.KEEPER:
      var player = game.players[game.activePlayer];
      player.keepers = player.keepers || [];
      player.keepers.push(card);
      game.players[game.activePlayer] = player;
      break;
    case TYPES.RULE:
      var rule = card.rule;
      if (rule.category === "Draw") game.constantRules.draw = rule.amount;
      else if (rule.category === "Play") game.constantRules.play = rule.amount;
      break;

    default:
      break;
  }
  game = checkForWinner(game);
  gameIsValid(game);
  return game;
}
