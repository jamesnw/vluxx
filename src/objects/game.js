import { cards } from "./cards";
import { dealCard } from "./player";
import _ from "lodash";
export default class Game {
  constructor() {
    this.id = "foo";
    this.players = [];
    this.rules = [];
  }
  addPlayer(player) {
    this.players.push(player);
  }
  addRule(rule) {
    this.rules.push(rule);
  }
}

export function newGame(name) {
  return {
    name: name,
    players: [],
    rules: []
  };
}

export function startGame(game) {
  game.deck = _.shuffle(cards);
  game.state = "started";
  game.constantRules = {
    draw: 1,
    play: 1
  };
  game.turn = {};
  // deal
  for (let index = 0; index < 3; index++) {
    game.players.forEach((player, index) => {
      dealCard(game, index);
    });
  }
  game.activePlayer = 0;
  return game;
}
