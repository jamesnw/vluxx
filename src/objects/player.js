import _ from "lodash";
import { applyCard } from "./cards";
import { gameIsValid } from "./game";

export function dealCard(game, index) {
  index = index === undefined ? game.activePlayer : index;
  const card = game.deck.shift();
  game.players[index].hand = game.players[index].hand || [];
  game.players[index].hand.push(card);
  gameIsValid(game);
  return game;
}

export function playCard(game, card) {
  game.discard = game.discard || [];
  game.discard.push(card);
  let hand = game.players[game.activePlayer].hand;
  _.remove(hand, { title: card.title });
  game.players[game.activePlayer].hand = hand;
  game.turn.playsRemaining = game.turn.playsRemaining - 1;
  game = applyCard(game, card);
  gameIsValid(game);
  return game;
}

export function startTurn(game) {
  const cardsToDraw = game.constantRules.draw;
  for (let index = 0; index < cardsToDraw; index++) {
    game = dealCard(game);
  }
  game.turn = {
    playsRemaining: game.constantRules.play
  };
  gameIsValid(game);
  return game;
}

export function endTurn(game) {
  game.turn = { playsRemaining: null };
  game.activePlayer = (game.activePlayer + 1) % game.players.length;
  gameIsValid(game);
  return game;
}
