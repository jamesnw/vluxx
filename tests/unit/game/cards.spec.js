import { applyCard } from "@/objects/cards";
import { startedGame } from "../fixtures/game";
import * as gameObject from "@/objects/game";

describe("cards", () => {
  let game;
  beforeEach(() => {
    game = startedGame();
  });
  describe("applyCard", () => {
    it("should set goal", () => {
      let card = {
        title: "Tools",
        type: "goal",
        goal: ["Hammer", "Screwdriver"]
      };
      game = applyCard(game, card);
      expect(game.goal).toEqual(card);
      expect(game).toMatchSnapshot();
    });
    it("should add keeper to player's keeper pile", () => {
      let card = { title: "Chocolate", type: "keeper" };
      game = applyCard(game, card);
      expect(game.players[1].keepers[0]).toEqual(card);
      expect(game).toMatchSnapshot();
    });
    describe("rules", () => {
      it("should apply new draw rule", () => {
        let card = {
          title: "Draw 2",
          type: "rule",
          rule: { category: "Draw", amount: 2 }
        };
        game = applyCard(game, card);
        expect(game.constantRules.draw).toEqual(2);
        expect(game).toMatchSnapshot();
      });
      it("should apply new play rule", () => {
        let card = {
          title: "Play 2",
          type: "rule",
          rule: { category: "Play", amount: 2 }
        };
        game = applyCard(game, card);
        expect(game.constantRules.play).toEqual(2);
        expect(game).toMatchSnapshot();
      });
    });
    it("should check game validity", () => {
      let gameIsValid = jest.spyOn(gameObject, "gameIsValid");
      applyCard(game, { type: "foo" });
      expect(gameIsValid).toHaveBeenCalled();
    });
    it("should check for winner", () => {
      let checkForWinner = jest.spyOn(gameObject, "checkForWinner");
      applyCard(game, { type: "foo" });
      expect(checkForWinner).toHaveBeenCalled();
    });
  });
});
