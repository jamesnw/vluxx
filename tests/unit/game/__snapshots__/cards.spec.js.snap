// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cards applyCard rules should apply new draw rule 1`] = `
Object {
  "activePlayer": 1,
  "constantRules": Object {
    "draw": 2,
    "play": 1,
  },
  "deck": Array [],
  "discard": Array [],
  "name": "Foo",
  "players": Array [
    Object {
      "hand": Array [],
      "name": "Jasper",
    },
    Object {
      "hand": Array [],
      "name": "Forde",
    },
  ],
  "rules": Array [],
  "state": "started",
  "turn": Object {
    "playsRemaining": 1,
  },
}
`;

exports[`cards applyCard rules should apply new play rule 1`] = `
Object {
  "activePlayer": 1,
  "constantRules": Object {
    "draw": 1,
    "play": 2,
  },
  "deck": Array [],
  "discard": Array [],
  "name": "Foo",
  "players": Array [
    Object {
      "hand": Array [],
      "name": "Jasper",
    },
    Object {
      "hand": Array [],
      "name": "Forde",
    },
  ],
  "rules": Array [],
  "state": "started",
  "turn": Object {
    "playsRemaining": 1,
  },
}
`;

exports[`cards applyCard should add keeper to player's keeper pile 1`] = `
Object {
  "activePlayer": 1,
  "constantRules": Object {
    "draw": 1,
    "play": 1,
  },
  "deck": Array [],
  "discard": Array [],
  "name": "Foo",
  "players": Array [
    Object {
      "hand": Array [],
      "name": "Jasper",
    },
    Object {
      "hand": Array [],
      "keepers": Array [
        Object {
          "title": "Chocolate",
          "type": "keeper",
        },
      ],
      "name": "Forde",
    },
  ],
  "rules": Array [],
  "state": "started",
  "turn": Object {
    "playsRemaining": 1,
  },
}
`;

exports[`cards applyCard should set goal 1`] = `
Object {
  "activePlayer": 1,
  "constantRules": Object {
    "draw": 1,
    "play": 1,
  },
  "deck": Array [],
  "discard": Array [],
  "goal": Object {
    "goal": Array [
      "Hammer",
      "Screwdriver",
    ],
    "title": "Tools",
    "type": "goal",
  },
  "name": "Foo",
  "players": Array [
    Object {
      "hand": Array [],
      "name": "Jasper",
    },
    Object {
      "hand": Array [],
      "name": "Forde",
    },
  ],
  "rules": Array [],
  "state": "started",
  "turn": Object {
    "playsRemaining": 1,
  },
}
`;
